assembly/src/tokens/tokenizer.rs

```diff
@@ -0,0 +1,169 @@
+use super::{LineInfo, SourceLocation, Token};
+use core::mem;
+
+// LINE TOKENIZER
+// ================================================================================================
+
+/// A line tokenizer that will generate tokens with their locations from [LinesStream] and
+/// [LineInfo].
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct LineTokenizer<'a> {
+    line: &'a str,
+    location: SourceLocation,
+    dangling: Option<SourceLocation>,
+}
+
+impl<'a> From<&LineInfo<'a>> for LineTokenizer<'a> {
+    fn from(info: &LineInfo<'a>) -> Self {
+        let line_number = info.line_number();
+        let column = info.char_offset() + 1;
+        let location = SourceLocation::new(line_number, column);
+        let line = info.contents().unwrap_or("");
+        Self {
+            line,
+            location,
+            dangling: None,
+        }
+    }
+}
+
+impl<'a> LineTokenizer<'a> {
+    // STATE MUTATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Takes dangling docs location, if present.
+    pub fn take_dangling(&mut self) -> Option<SourceLocation> {
+        self.dangling.take()
+    }
+}
+
+impl<'a> Iterator for LineTokenizer<'a> {
+    type Item = (&'a str, SourceLocation);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.line.is_empty() {
+            return None;
+        }
+
+        if self.line.starts_with(Token::DOC_COMMENT_PREFIX) {
+            let mut location = self.location;
+            location.move_column(1);
+            self.dangling.replace(location);
+            return None;
+        }
+
+        if self.line.starts_with(Token::COMMENT_PREFIX) {
+            return None;
+        }
+
+        // [LinesStream] generates [LineInfo] without leading whitespaces
+        let (token, remainder) = match self.line.split_once(char::is_whitespace) {
+            Some(split) => split,
+            None => {
+                let line = mem::take(&mut self.line);
+                let location = mem::take(&mut self.location);
+                return Some((line, location));
+            }
+        };
+
+        // set location & find next non-whitespace
+        let location = self.location;
+        let at = match remainder.find(|c: char| !c.is_whitespace()) {
+            Some(at) => at,
+            None => {
+                mem::take(&mut self.line);
+                mem::take(&mut self.location);
+                return Some((token, location));
+            }
+        };
+
+        // update location, line & return
+        self.location.move_column((token.len() + at + 1) as u32);
+        self.line = remainder.split_at(at).1;
+        Some((token, location))
```

Comment:
We can probably simplify this slightly as follows:
```Rust
let token_loc = self.location;
let (token, offset) = match self.line.split_once(char::is_whitespace) {
    Some((token, remainder)) => {
        let offset = remainder.find(|c: char| !c.is_whitespace()).unwrap_or_default();
        (token, token.len() + offset + 1)
    }
    None => (self.line, self.line.len()),
};

let (_, remainder) = self.line.split_at(offset);
self.line = remainder;
self.location.move_column(offset as u32);

Some((token, token_loc))
```

Reply:
I preserved the previous checks & update it here: https://github.com/0xPolygonMiden/miden-vm/pull/861/commits/164a49d0ad0dcbb3b390f8e47d2f2c5d4df121ed

Reply:
Yes - my suggestion was to replace only the suggested lines and keep the checks for comments. I believe all tests are passing now - right?